### ISSUE TITLE
Memory leak and wrong behavior in fluid.layers.cross_entropy

### ISSUE THREAD

**Issue by Officium** [PaddlePaddle/Paddle#17232]
**System information**
-PaddlePaddle version: 1.4
-CPU: No MKL or OpenBLAS
-GPU: No GPU
-OS Platform: Mac OS 10.14.4
**To Reproduce**
```python
import numpy as np
import paddle.fluid as fluid


data = fluid.layers.data(name='data', shape=[5])
label = fluid.layers.data(name='label', shape=[1])
predict = fluid.layers.fc(name='fc', input=data, size=1, act='sigmoid')
cost = fluid.layers.cross_entropy(predict, label, soft_label=False)

place = fluid.CPUPlace()
exe = fluid.Executor(place)
exe.run(fluid.default_startup_program())
x = np.random.random((10, 5)).astype('float32')
y = np.random.randint(0, 2, (10, 1)).astype('int64')
result = exe.run(fluid.default_main_program(),
                 feed={'data': x,
                       'label': y},
                 fetch_list=[cost],
                 return_numpy=True)
print(x)
print(y)
print(result)
```
**Describe your current behavior**
No data shape check in `fluid.layers.cross_entropy` and sometimes return very big loss like `1e20`.
**One sample behavior**
```python
[[0.49041983 0.7517153  0.97184145 0.53451246 0.32012275]
 [0.01281779 0.82265323 0.9833126  0.28373203 0.12450393]
 [0.8151463  0.52735615 0.80351907 0.22486255 0.594022  ]
 [0.588412   0.8489567  0.43820578 0.8960106  0.8428685 ]
 [0.5137817  0.26190907 0.22895095 0.23925517 0.21312556]
 [0.25527868 0.7975383  0.70617896 0.0762081  0.12858906]
 [0.01175127 0.19969502 0.14499383 0.2757372  0.07508669]
 [0.5904928  0.36498263 0.89440274 0.29967937 0.5139649 ]
 [0.8936788  0.29070398 0.03116068 0.41256332 0.70924914]
 [0.6440318  0.7952011  0.07752277 0.4691725  0.4938896 ]]
[[1]
 [1]
 [0]
 [0]
 [1]
 [1]
 [1]
 [0]
 [0]
 [1]]
[array([[1.2810558e+00],
       [1.1927774e+00],
       [1.1927774e+00],
       [1.3012718e+00],
       [1.2652258e+00],
       [8.2041526e-01],
       [1.0734150e+00],
       [1.0734150e+00],
       [9.6498191e-01],
       [1.0000000e+20]], dtype=float32)]
```

**Comment by chengduoZH** [comment]
@Officium Can you describe how you observed memory leaks?

**Comment by Officium** [comment]
@chengduoZH Please try the following code. The memory usage will increase slowly. And when set `n = 1e8` the script will raise `Segmentation fault: 11`. 
Another observation is that larger batch size will not speed up the memory increasing. So maybe there is only index out of bounds and no memory leak.

```python
import numpy as np
import paddle.fluid as fluid
from tqdm import trange


data = fluid.layers.data(name='data', shape=[5])
label = fluid.layers.data(name='label', shape=[1])
predict = fluid.layers.fc(name='fc', input=data, size=1, act='sigmoid')
cost = fluid.layers.cross_entropy(predict, label, soft_label=False)
avg_cost = fluid.layers.mean(cost)
optimizer = fluid.optimizer.SGD(learning_rate=5e-4)
optimizer.minimize(avg_cost)

place = fluid.CPUPlace()
exe = fluid.Executor(place)
exe.run(fluid.default_startup_program())

batch_size = 1
n = 1
x = np.random.random((batch_size, 5)).astype('float32')
y = (np.zeros((batch_size, 1)) + n).astype('int64')
for _ in trange(10000000):
    result = exe.run(fluid.default_main_program(),
                     feed={'data': x, 'label': y},
                     fetch_list=[avg_cost], return_numpy=True)

```

**Comment by chengduoZH** [comment]
@Officium Thanks for your reply. I have fixed this bug in #17282.
